validator/inputs/writer.go

```diff
@@ -0,0 +1,129 @@
+package inputs
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/offchainlabs/nitro/validator/server_api"
+)
+
+// Writer is a configurable writer of InputJSON files.
+//
+// The default Writer will write to a path like:
+//
+//	$HOME/.arbuitrum/validation-inputs/<YYYMMDD_HHMMSS>/block_inputs_<id>.json
+//
+// The path can be nested under a slug directory so callers can provide a
+// recognizable name to differentiate various contexts in which the InputJSON
+// is being written. If the Writer is configured by calling SetSlug, then the
+// path will be like:
+//
+//	$HOME/.arbuitrum/validation-inputs/<slug>/<YYYMMDD_HHMMSS>/block_inputs_<id>.json
+//
+// The inclusion of a timestamp directory is on by default to avoid conflicts which
+// would result in files being overwritten. However, the Writer can be configured
+// to not use a timestamp directory.  If the Writer is configured by calling
+// SetUseTimestampDir(false), then the path will be like:
+//
+//	$HOME/.arbuitrum/validation-inputs/<slug>/block_inputs_<id>.json
+//
+// Finally, to give complete control to the clients, the base directory can be
+// set directly with SetBaseDir. In which case, the path will be like:
+//
+//	<baseDir>/block_inputs_<id>.json
+//	  or
+//	<baseDir>/<slug>/block_inputs_<id>.json
+//	  or
+//	<baseDir>/<slug>/<YYYMMDD_HHMMSS>/block_inputs_<id>.json
+type Writer struct {
+	clock           Clock
+	baseDir         string
+	slug            string
+	useTimestampDir bool
+}
+
+// Clock is an interface for getting the current time.
+type Clock interface {
+	Now() time.Time
+}
+
+type realClock struct{}
+
+func (realClock) Now() time.Time {
+	return time.Now()
+}
+
+// NewWriter creates a new Writer with default settings.
+func NewWriter() (*Writer, error) {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return nil, err
+	}
+	baseDir := fmt.Sprintf("%s/.arbitrum/validation-inputs", homeDir)
```

Comment:
not sure how easy it is to get here.. but a good base dir would be stack.InstanceDir()

Reply:
It looks slightly odd to wire together. But, I suggest that good separation of concerns would keep the inputs.Writer from knowing about the stack at all. If anything, I could see the various clients of this library (in the common case the `BlockValidator`) getting the `stack.InstanceDir()` and passing in using the `WithBaseDir` option.

... but, even the BlockValidator doesn't currently have convienent access to the `stack` directly.
`NewStatelessBlockValidator` is currently being passed a `stack` argument. But, it doesn't store the reference to it anywhere.
If you feel strongly that we should pass in `stack.InstanceDir()` from the BlockValidator, we could add a reference to the stack as a member of the `StatelessBlockValidator` and then use it to come up with the `InstanceDir()`

In fact, I'll just go ahead and wire that together now, and I can back it out if you think it's getting too messy.

Reply:
I agree it's not very pretty - but the good thing about stack.InstanceDir is that we know this process has write permission there, and that this directory is saved in reboot.
I think we've never had real failure. If we by default write to a dir that doesn't have these two properties - we probably wouldn't know until it's too late.